File: modules/widgets/lib/aerox/widgets/containers/Sizer.cpp
#include "aerox/widgets/containers/Sizer.hpp"

#include "aerox/widgets/ContainerSlot.hpp"

namespace aerox::widgets
{
    void Sizer::SetWidthOverride(const std::optional<float>& width)
    {
        _widthOverride = width;
        CheckSize();
    }

    void Sizer::SetHeightOverride(const std::optional<float>& height)
    {
        _heightOverride = height;
        CheckSize();
    }

    Vec2<float> Sizer::ComputeDesiredSize()
    {
        if(const auto slot = GetSlot(0))
        {
            auto size = slot->GetWidget()->GetDrawSize();
            return Vec2{_widthOverride.value_or(size.x),_heightOverride.value_or(size.y)};
        }
        
        return Vec2{_widthOverride.value_or(0),_heightOverride.value_or(0)};
    }

    size_t Sizer::GetMaxSlots() const
    {
        return 1;
    }

    void Sizer::ArrangeSlots(const Vec2<float>& drawSize)
    {
        if(const auto slot = GetSlot(0))
        {
            auto widget = slot->GetWidget();
            widget->SetRelativeOffset(Vec2{0.0f,0.0f});
            auto padding = GetPadding();
            widget->SetDrawSize(GetDesiredSize() - Vec2{padding.left + padding.right,padding.top + padding.bottom});
        }
    }
}

File: aerox.Runtime.Widgets/Containers/Sizer.cs
using aerox.Runtime.Math;

namespace aerox.Runtime.Widgets.Containers;

public class Sizer : Container
{
    private float? _heightOverride;
    private float? _widthOverride;

    public Sizer(Widget child) : base([child])
    {
    }

    public Sizer()
    {
    }

    public float? WidthOverride
    {
        get => _widthOverride;
        set
        {
            _widthOverride = value;
            CheckSize();
        }
    }

    public float? HeightOverride
    {
        get => _heightOverride;
        set
        {
            _heightOverride = value;
            CheckSize();
        }
    }

    protected override Size2d ComputeContentDesiredSize()
    {
        if (GetSlot(0) is { } slot)
        {
            var desiredSize = slot.GetWidget().GetDesiredSize();
            return new Size2d(WidthOverride ?? desiredSize.Width, HeightOverride ?? desiredSize.Height);
        }

        return new Size2d();
    }

    public override uint GetMaxSlots() => 1;

    protected override void ArrangeSlots(Size2d drawSize)
    {
        if (GetSlot(0) is { } slot)
        {
            slot.GetWidget().SetOffset(new Vector2<float>(0, 0));
            slot.GetWidget().SetSize(drawSize.Clone());
        }
    }
}